### PR TITLE
validate rma lane if rkey is not needed for mem type

### LINKED ISSUE #2919: Segfault in CUDA support when KNEM not present and CMA is active intra-node RMA transport
When KNEM is not present, CMA will be selected for RMA for which RKEY is not needed. When RKEY is not needed, it is wrongly selecting CMA lane for cuda mem type rma.

Workarounds:
    install KNEM
    (or)
    Disable CMA  (--disable-cma)
    (or)
    not select cma through UCX_TLS.

### PR DESCRIPTION
**Pull request by bureddy**
## What
Fix rma lane selection for CUDA mem type when there is lane with is not required RKEY(ex CMA)

## Why ?
When KNEM is not present, CMA will be selected for RMA for which RKEY is not needed. When RKEY is not needed, it is hard determinically indicate if it can do RMA to specific mem type. With current mem type protocols it is not possible to support all combinations for src and dst mem type combinations( {(host,host) (host,cuda), (cuda,host), (cuda, cuda)) in this scenario.

Fixes #2919

## How ?
pack mem_type info in RKEY and validate rma lane mem type with  remote and local memory type

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8104/ for details (Mellanox internal link).

File: src/ucp/core/ucp_rkey.c [hunk 1]
```diff
@@ -40,6 +42,8 @@ void ucp_rkey_packed_copy(ucp_context_h context, ucp_md_map_t md_map,
     *(ucp_md_map_t*)p = md_map;
     p += sizeof(ucp_md_map_t);
 
+    *((uint8_t *)p++) = mem_type;
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_rkey.c]: pls use `uct_memory_type_t`  instead of `uint8_t` for casting and `sizeof` where relevant

Comment (bureddy) [reply src/ucp/core/ucp_rkey.c]: @evgeny-leksikov uct_memory_type_t is defined as enum.  wanted to save on rkey packed size with typecasting to uint8_t.

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_rkey.c]: Ok, then pls check if we have static assert for UCT_MD_MEM_TYPE_LAST <= 255 or similar

Comment (bureddy) [reply src/ucp/core/ucp_rkey.c]: added assert while packing rkey

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8134/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8141/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8142/ for details (Mellanox internal link).

File: src/ucp/core/ucp_rkey.c [hunk 2]
```diff
@@ -13,7 +13,10 @@
 #include <inttypes.h>
 
 
-static ucp_md_map_t ucp_mem_dummy_buffer = 0;
+static struct {
+    ucp_md_map_t md_map;
```
Comment (yosefe) [inline src/ucp/core/ucp_rkey.c]: indent on column

File: src/ucp/core/ucp_rkey.c [hunk 3]
```diff
@@ -13,7 +13,10 @@
...
+    uint8_t mem_type;
+} UCS_S_PACKED ucp_mem_dummy_buffer = {0, 0};
```
Comment (yosefe) [inline src/ucp/core/ucp_rkey.c]: maybe use UCT_MEM_TYPE_xx constant instead of 0?

File: src/ucp/core/ucp_rkey.c [hunk 4]
```diff
@@ -349,16 +370,20 @@ static ucp_lane_index_t ucp_config_find_rma_lane(ucp_context_h context,
         }
 
         md_index = config->md_index[lane];
+        md_attr  = &context->tl_mds[md_index].attr;
+
         if ((md_index != UCP_NULL_RESOURCE) &&
-            (!(context->tl_mds[md_index].attr.cap.flags & UCT_MD_FLAG_NEED_RKEY)))
+            (!(md_attr->cap.flags & UCT_MD_FLAG_NEED_RKEY)) &&
+            (!rkey || (md_attr->cap.mem_type == mem_type &&
```
Comment (yosefe) [inline src/ucp/core/ucp_rkey.c]: ( ) around internal conditions
also - maybe use helper bool variable? it's hard to parse this..

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8158/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8160/ for details (Mellanox internal link).

File: src/ucp/core/ucp_rkey.c [hunk 5]
```diff
@@ -349,16 +370,21 @@ static ucp_lane_index_t ucp_config_find_rma_lane(ucp_context_h context,
...
+        md_attr  = &context->tl_mds[md_index].attr;
+
         if ((md_index != UCP_NULL_RESOURCE) &&
-            (!(context->tl_mds[md_index].attr.cap.flags & UCT_MD_FLAG_NEED_RKEY)))
+            (!(md_attr->cap.flags & UCT_MD_FLAG_NEED_RKEY)))
         {
             /* Lane does not need rkey, can use the lane with invalid rkey  */
-            *uct_rkey_p = UCT_INVALID_RKEY;
-            return lane;
+            if (!rkey || ((md_attr->cap.mem_type == mem_type) &&
```
[the remark this answers is not in the document]
Comment (bureddy) [reply src/ucp/core/ucp_rkey.c]: currently,  RKEY is not needed for pipeline stating protocol on local mem type(cuda) endpoint rma bw lane (only cuda_copy)  for it's zcopy ops to move data from GPU to CPU memory. So, instead creating dummy key pack/unpack I had passed passed NULL  here (https://github.com/openucx/ucx/pull/2995/files#diff-f89c8c0721aa5b2b55347e1612b8395bR880).
